pennylane/beta/interfaces/tf.py

```diff
@@ -0,0 +1,187 @@
+
+    The TensorFlow interface defaults to ``tf.float64`` output. This can be modified by
+    providing the ``dtype`` argument when applying the interface:
+
+    >>> p = tf.Variable([0.1, 0.2, 0.3], dtype=tf.float32)
+    >>> with tf.GradientTape() as tape:
+    ...     TFInterface.apply(qtape, dtype=tf.float32)  # reusing the previous qtape
+    ...     result = qtape.execute(dev)
+    >>> print(result)
+    tf.Tensor([0.06982072], shape=(1,), dtype=float32)
+    >>> grad = tape.gradient(result, p)
+    >>> print(grad)
+    tf.Tensor([0.2895088  0.38464668 0.09645163], shape=(3,), dtype=float32)
+    """
+
+    dtype = tf.float64
+
+    @property
+    def interface(self):  # pylint: disable=missing-function-docstring
+        return "tf"
+
+    def _update_trainable_params(self):
+        params = self.get_parameters(trainable_only=False)
+
+        trainable_params = set()
+
+        for idx, p in enumerate(params):
+            # Determine which input tensors/Variables are being recorded for backpropagation.
+            # The function should_record_backprop, documented here:
+            # https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/eager/tape.py#L163
+            # accepts lists of *tensors* (not Variables), returning True if all are being watched by one or more
+            # existing gradient tape, False if not.
+            if isinstance(p, (tf.Variable, tf.Tensor)) and should_record_backprop(
+                [tf.convert_to_tensor(p)]
+            ):
+                trainable_params.add(idx)
+
+        self.trainable_params = trainable_params
+
+    @staticmethod
+    def convert_to_numpy(tensors):
+        """Converts any TensorFlow tensors in a sequence to NumPy arrays.
+
+        Args:
+            tensors (Sequence[Any, tf.Variable, tf.Tensor]): input sequence
+
+        Returns:
+            list[Any, array]: list with all tensors converted to NumPy arrays
+        """
+        return [i.numpy() if isinstance(i, (tf.Variable, tf.Tensor)) else i for i in tensors]
+
+    @tf.custom_gradient
+    def _execute(self, params, **input_kwargs):
+        # unwrap free parameters
+        args = self.convert_to_numpy(params)
+
+        # unwrap constant parameters
+        all_params = self.get_parameters(trainable_only=False)
+        all_params_unwrapped = self.convert_to_numpy(all_params)
+
+        self.set_parameters(all_params_unwrapped, trainable_only=False)
+        res = self.execute_device(args, input_kwargs["device"])
+        self.set_parameters(all_params, trainable_only=False)
+
+        def grad(grad_output, **tfkwargs):
+            variables = tfkwargs.get("variables", None)
+
+            self.set_parameters(all_params_unwrapped, trainable_only=False)
+            jacobian = self.jacobian(input_kwargs["device"], params=args, **self.jacobian_options)
+            self.set_parameters(all_params, trainable_only=False)
+
+            jacobian = tf.constant(jacobian, dtype=self.dtype)

```

Comment:
How come this would be needed?

Reply:
This is because the output of `self.jacobian` is a NumPy tensor. So we need to convert it to a TensorFlow tensor, to allow it to be used in matrix multiplication below